thetis/shallowwater_eq.py

```diff
@@ -342,6 +346,65 @@ def rhs_implicit(self, solution, wind_stress=None,
 
         return -f - g
 
+    def horizontal_viscosity(self, uv, nu, total_h):
+
+        n = FacetNormal(self.mesh)
+
+        if self.include_grad_div_viscosity_term:
+            stress = nu*2.*sym(grad(uv))
+            stress_jump = avg(nu)*2.*sym(tensor_jump(uv, n))
+        else:
+            stress = nu*grad(uv)
+            stress_jump = avg(nu)*tensor_jump(uv, n)
+
+        f = inner(grad(self.U_test), stress)*dx
+
+        if self.u_is_dg:
+            # projection to DG0 is only necessary because of a bug that's now fixed on master firedrake:
+            h = project(CellSize(self.mesh), FunctionSpace(self.mesh, "DG", 0))
+            # from Epshteyn et al. 2007 (http://dx.doi.org/10.1016/j.cam.2006.08.029)
+            # the scheme is stable for alpha > 3*X*p*(p+1)*cot(theta), where X is the
+            # maximum ratio of viscosity within a triangle, p the degree, and theta
+            # with X=2, theta=6: cot(theta)~10, 3*X*cot(theta)~60
+            p = self.U_space.ufl_element().degree()
+            alpha = 60.*p*(p+1)
```

Comment:
I've used a smaller value `alpha = 3.5*p*(p+1)`. I presume that some of your test cases require such a high penalty?